onnxruntime/core/optimizer/nhwc_transformer.cc

```diff
@@ -0,0 +1,366 @@
+  if (parent_node) {
+    graph_.AddEdge(parent_node->Index(), permute_node.Index(), srcIdx, 0);
+  }
+
+  return permute_node.Index();
+}
+
+NodeIndex NhwcTransformerImpl::InsertPermuteChildNode(Node& node, bool bNHWC) {
+
+  LOGS(logger, VERBOSE) << "NHWC Insert: " << node.OpType()
+    << (bNHWC ? " [ReorderInput] " : " [ReorderOutput] ") << "*:0";
+
+  auto& output_defs = node.MutableOutputDefs();
+  std::string new_output_def_name = graph_.GenerateNodeArgName("output");
+  auto* new_output_arg = &graph_.GetOrCreateNodeArg(new_output_def_name, output_defs[0]->TypeAsProto());
+
+  Node& permute_node = graph_.AddNode(graph_.GenerateNodeName(bNHWC ? "PermuteNCHW" : "PermuteNHWC"),
+                                     bNHWC ? "ReorderIntput" : "ReorderOutput",
+                                     bNHWC ? "ReorderIntput" : "ReorderOutput",
+                                     {new_output_arg},
+                                     {output_defs[0]},
+                                     nullptr,
+                                     kMSNhwcDomain);
+  permute_node.SetExecutionProviderType(provider_);
+  output_defs[0] = new_output_arg;
+  graph_.AddEdge(node.Index(), permute_node.Index(), 0, 0);
+
+  return permute_node.Index();
+}
+
+bool NhwcTransformerImpl::RequiresWeightsPermutation(const Node& node) {
+   return ((node.GetExecutionProviderType() == kAclExecutionProvider ||
+            node.GetExecutionProviderType() == kArmNNExecutionProvider) &&
+           (node.OpType() == "Conv" ||
+            node.OpType() == "FusedConv"));
+}
+
+NodeIndex NhwcTransformerImpl::ReplaceNode(Node& node) {
+
+  auto& input_defs = node.MutableInputDefs();
+  auto& output_defs = node.MutableOutputDefs();
+
+  LOGS(logger, VERBOSE) << "NHWC >> Replace " << node.OpType();
+
+  Node& newNode = graph_.AddNode(graph_.GenerateNodeName(node.Name()),
+                                     node.OpType(),
+                                     node.OpType(),
+                                     input_defs,
+                                     output_defs,
+                                     &node.GetAttributes(),
+                                     kMSNhwcDomain);
+  newNode.SetExecutionProviderType(node.GetExecutionProviderType());
+
+  // Permute weights
+  if (RequiresWeightsPermutation(node))
+    PermuteWeights(input_defs[1], &newNode.MutableInputDefs()[1], node.GetExecutionProviderType());
+
+  NodeIndex oldIndex = node.Index();
+  const std::vector<std::reference_wrapper<Node>> replacedNode({node});
+  graph_utils::FinalizeNodeFusion(graph_, replacedNode, newNode);
+  nodes_layout.erase(oldIndex);
+
+  return newNode.Index();
+}
+
+void NhwcTransformerImpl::PermuteWeights(NodeArg *input_def, NodeArg** nhwc_conv_W_arg, __attribute__ ((unused)) const std::string& execution_provider) {
+  // Require that the weights tensor be static.
+  const ONNX_NAMESPACE::TensorProto* conv_W_tensor_proto = nullptr;
```

Comment:
Please use graph_utils::IsConstantInitializer to check if they're constant. Initializers can potentially be overridden at runtime if a matching name is a graph input.

Reply:
Some models from ONNX Zoo like squeezenet does not mark convolution weights as constants and ONNX convolution schema does not seem to enforce this. I suppose these models do not work properly with NCHWC transformation either.

Reply:
If the model IR version is < 4 all initializers get treated as constants. I would expect older models in the model zoo are from before IR v4 (e.g. [https://github.com/onnx/models/blob/master/vision/classification/squeezenet/model/squeezenet1.1-7.onnx](https://github.com/onnx/models/blob/master/vision/classification/squeezenet/model/squeezenet1.1-7.onnx) is IR v3). graph_utils::IsConstantInitializer takes into account the IR version and also handles subgraphs where the initializer may come from a parent graph.

---
In reply to: [441385457](https://github.com/microsoft/onnxruntime/pull/4237#discussion_r441385457) [](ancestors = 441385457)